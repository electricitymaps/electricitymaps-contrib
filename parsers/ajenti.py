"""
Initial PR https://github.com/electricitymaps/electricitymaps-contrib/pull/2456
Discussion thread https://github.com/electricitymaps/electricitymaps-contrib/issues/636
A promotion webpage for King's Island energy production is here: https://www.hydro.com.au/clean-energy/hybrid-energy-solutions/success-stories/king-island
As of 09/2020, it embeds with <iframe> the URI https://data.ajenti.com.au/KIREIP/index.html

About the data, the feed we get seems to be counters with a 2 seconds interval.
That means that if we fetch these counters every 15 minutes, we only are reading "instantaneous" metters that could differ from the total quantity of energies at play.
To get the very exact data, we would need to have a parser running constanty to collect those 2-sec interval counters.
"""

import json
from datetime import datetime
from logging import Logger, getLogger
from zoneinfo import ZoneInfo

from requests import Session
from signalr import Connection

ZONE_PARAMS = {
    # King Island
    # https://www.hydro.com.au/clean-energy/hybrid-energy-solutions/success-stories/king-island
    "AU-TAS-KI": {
        "hub": "TagHub",
        "method": "Dashboard",
        "tz": "Australia/Currie",
        "source": "hydro.com.au",  # Iframe: https://data.ajenti.com.au/KIREIP/index.html
    },
    # Flinders Island
    # https://github.com/electricitymaps/electricitymaps-contrib/issues/2533
    # https://en.wikipedia.org/wiki/Flinders_Island
    # https://www.hydro.com.au/clean-energy/hybrid-energy-solutions/success-stories/flinders-island
    "AU-TAS-FI": {
        "hub": "flindershub",
        "method": "SendDashboard",
        "tz": "Australia/Hobart",
        "source": "hydro.com.au",
    },
    # Rottnest Island
    # https://github.com/electricitymaps/electricitymaps-contrib/issues/2534
    # https://en.wikipedia.org/wiki/Rottnest_Island
    # https://www.hydro.com.au/clean-energy/hybrid-energy-solutions/success-stories/rottnest-island
    "AU-WA-RI": {
        "hub": "HogsHub",
        "method": "SendDashboard",
        "tz": "Australia/Perth",
        "source": "hydro.com.au",
    },
}


class SignalR:
    def __init__(self, url):
        self.url = url

    def update_res(self, msg):
        if msg != {}:
            self.res = msg

    def get_value(self, hub, method):
        self.res = {}
        with Session() as session:
            # create a connection
            connection = Connection(self.url, session)
            chat = connection.register_hub(hub)
            chat.client.on(method, self.update_res)
            connection.start()
            connection.wait(3)
            connection.close()
            return self.res


def parse_payload(logger: Logger, payload) -> dict:
    technologies_parsed = {
        "biomass": 0,
        "battery": 0,
        "coal": 0,
        "flywheel": 0,
        "gas": 0,
        "hydro": 0,
        "nuclear": 0,
        "oil": 0,
        "solar": 0,
        "wind": 0,
        "geothermal": 0,
        "unknown": 0,
    }
    if "technologies" not in payload:
        raise KeyError(
            f"No 'technologies' in payload\n" f"serie : {json.dumps(payload)}"
        )
    else:
        logger.debug(f"serie : {json.dumps(payload)}")
    for technology in payload["technologies"]:
        # rename upstream key to match our payload needs
        if technology["id"] == "diesel":
            technology["id"] = "oil"

        assert technology["unit"] == "kW"
        # The upstream API gives us kW, we need MW
        technologies_parsed[technology["id"]] = int(technology["value"]) / 1000

    # King's Island uses some percentage of biodiesel; since we account differently traditionnal oil & biomass, here we do math to separate both
    if "biodiesel" in payload:
        # sanity check to be sure that we are not remaking data that the data source could otherwise give us
        assert "biomass" and "oil" != 0
        technologies_parsed["biomass"] = (
            technologies_parsed["oil"] * payload["biodiesel"]["percent"] / 100
        )
        technologies_parsed["oil"] = (
            technologies_parsed["oil"] * (100 - payload["biodiesel"]["percent"]) / 100
        )

    logger.debug(f"production : {json.dumps(technologies_parsed)}")

    return technologies_parsed


# Both keys battery and flywheel are negative when storing energy, and positive when feeding energy to the grid
def format_storage_techs(technologies_parsed):
    # sometimes we don't get those keys. For instance Rottnest island don't have any storage.
    if "battery" and "flywheel" in technologies_parsed:
        storage_techs = technologies_parsed["battery"] + technologies_parsed["flywheel"]
    else:
        storage_techs = 0
    _battery_production = (
        storage_techs if storage_techs > 0 else 0
    )  # TODO: Should this just be removed?
    _battery_storage = storage_techs if storage_techs < 0 else 0


def sum_storage_techs(technologies_parsed):
    storage_techs = technologies_parsed["battery"] + technologies_parsed["flywheel"]

    return storage_techs


def fetch_production(
    zone_key: str = "AU-TAS-KI",
    session: Session | None = None,
    target_datetime=None,
    logger: Logger = getLogger(__name__),
) -> dict:
    if target_datetime is not None:
        raise NotImplementedError(
            "The datasource currently implemented is only real time"
        )

    # get the specific zone parameters
    try:
        hub, dashboard, tz, source = (
            ZONE_PARAMS[zone_key]["hub"],
            ZONE_PARAMS[zone_key]["method"],
            ZONE_PARAMS[zone_key]["tz"],
            ZONE_PARAMS[zone_key]["source"],
        )
    except KeyError as e:
        raise KeyError("The zone " + zone_key + " isn't implemented") from e

    payload = SignalR("https://data.ajenti.com.au/live/signalr").get_value(
        hub, dashboard
    )
    technologies_parsed = parse_payload(logger, payload)
    storage_techs = sum_storage_techs(technologies_parsed)

    return {
        "zoneKey": zone_key,
        "datetime": datetime.now(tz=ZoneInfo(tz)),
        "production": {
            "biomass": technologies_parsed["biomass"],
            "coal": technologies_parsed["coal"],
            "gas": technologies_parsed["gas"],
            "hydro": technologies_parsed["hydro"],
            "nuclear": technologies_parsed["nuclear"],
            "oil": technologies_parsed["oil"],
            "solar": technologies_parsed["solar"],
            "wind": 0
            if technologies_parsed["wind"] < 0 and technologies_parsed["wind"] > -0.1
            else technologies_parsed[
                "wind"
            ],  # If wind between 0 and -0.1 set to 0 to ignore self-consumption
            "geothermal": technologies_parsed["geothermal"],
            "unknown": technologies_parsed["unknown"],
        },
        "storage": {
            "battery": storage_techs
            * -1  # Somewhat counterintuitively,to ElectricityMap positive means charging and negative means discharging
        },
        "source": source,
    }


if __name__ == "__main__":
    print(fetch_production())
