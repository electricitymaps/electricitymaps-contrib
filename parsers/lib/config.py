import functools
import os
from copy import deepcopy
from datetime import timedelta
from inspect import signature
from logging import getLogger

import requests
from requests import Session
from requests.adapters import HTTPAdapter, Retry


def refetch_frequency(frequency: timedelta):
    """Specifies the refetch frequency of a parser.
    The refetch frequency is used to determine the how much data is returned by the parser.

    i.e. if we refetch from d1 to d2 and the frequency is timedelta(days=1), then we will only
    call the function once every day between d1 and d2.
    """
    assert isinstance(frequency, timedelta)

    def wrap(f):
        def wrapped_f(*args, **kwargs):
            result = f(*args, **kwargs)
            return result

        wrapped_f.REFETCH_FREQUENCY = frequency
        return wrapped_f

    return wrap


def retry_policy(retry_policy: Retry):
    assert isinstance(retry_policy, Retry)

    def wrap(f):
        def wrapped_f(*args, **kwargs):
            session = args[1] if len(args) > 2 else kwargs.get("session")
            logger = kwargs.get("logger", getLogger(__name__))
            session = Session() if session is None else session
            old_adapters = None
            if "https://" in session.adapters or "http://" in session.adapters:
                logger.warning("Session already has adapters, they will be overriden.")
                old_adapters = deepcopy(session.adapters)
            session.mount("https://", HTTPAdapter(max_retries=retry_policy))
            session.mount("http://", HTTPAdapter(max_retries=retry_policy))
            result = f(*args, **kwargs)
            del session.adapters["https://"]
            del session.adapters["http://"]
            if old_adapters is not None:
                session.adapters.update(old_adapters)
            return result

        return wrapped_f

    return wrap


def use_proxy(country_code: str, monkeypatch_for_pydataxm: bool = False):
    """
    Decorator to route requests through webshare.io proxies for a specific country.

    This decorator should be used as a last resort when the data provider is non-responsive
    and other proxy methods (like cloud run datacenter proxies) do not work. Note that this
    proxy service is not free and is charged per GB.

    If you create an account with webshare.io to develop parsers for this
    project, you should subscribe to the "residential" proxy type (not "static
    residential"), as this choice includes most countries. As a practical
    example, South Korea proxies are only available using this subscription.

    Args:
        country_code (str):
            The ISO 3166-1 alpha-2 code of the country for which the proxy
            should be used.
        monkeypatch_for_pydataxm (bool):
            The CO parser specifically makes use both requests.post and
            aiohttp.ClientSession via a API specific third party library called
            pydataxm that isn't configurable to use a proxy. By setting this to
            True, we temporarily monkeypatch whats needed to get setup the
            proxy.

            See

    Usage:
    ```
        @use_proxy(country_code='MY')
        def fetch_production(zone_key, session, target_datetime, logger):
            ...
    ```
    """

    def wrap(f):
        sig = signature(f)
        if "zone_key" in sig.parameters:
            exchange_signature = False
        elif "zone_key1" in sig.parameters and "zone_key2" in sig.parameters:
            exchange_signature = True
        else:
            raise ValueError(
                "Invalid function signature. Maybe you added the @decorators in "
                "the wrong order? The use_proxy decorator should be the bottom decorator."
            )

        def wrapped_f(*args, **kwargs):
            WEBSHARE_USERNAME = os.environ.get("WEBSHARE_USERNAME")
            WEBSHARE_PASSWORD = os.environ.get("WEBSHARE_PASSWORD")
            if not WEBSHARE_USERNAME or not WEBSHARE_PASSWORD:
                logger = kwargs.get("logger", getLogger(__name__))
                logger.warning(
                    "Proxy environment variables are not set. "
                    "Attempting without proxy...\n"
                    "Add WEBSHARE_USERNAME and WEBSHARE_PASSWORD to use the proxy."
                )
                return f(*args, **kwargs)

            # get an existing Session object from args or kwargs, or create a
            # new one, so it can be temporarily re-configured
            if exchange_signature and len(args) >= 3:
                session = args[2]
            elif not exchange_signature and len(args) >= 2:
                session = args[1]
            else:
                session = kwargs.setdefault("session", Session())

            requests_proxy_dict = {
                "http": f"http://{WEBSHARE_USERNAME}-{country_code}-rotate:{WEBSHARE_PASSWORD}@p.webshare.io:80/",
                "https": f"http://{WEBSHARE_USERNAME}-{country_code}-rotate:{WEBSHARE_PASSWORD}@p.webshare.io:80/",
            }

            if monkeypatch_for_pydataxm:
                from aiohttp import BasicAuth, ClientSession

                aiohttp_old_post = ClientSession.post
                requests_old_post = requests.post

                ClientSession.post = functools.partialmethod(
                    ClientSession.post,
                    proxy="http://p.webshare.io",
                    proxy_auth=BasicAuth(
                        f"{WEBSHARE_USERNAME}-{country_code}-rotate",
                        WEBSHARE_PASSWORD,
                    ),
                )
                requests.post = functools.partial(
                    requests.post,
                    proxies=requests_proxy_dict,
                )

            old_proxies = session.proxies
            session.proxies = requests_proxy_dict
            try:
                return f(*args, **kwargs)
            finally:
                session.proxies = old_proxies
                if monkeypatch_for_pydataxm:
                    ClientSession.post = aiohttp_old_post
                    requests.post = requests_old_post

        return wrapped_f

    return wrap
