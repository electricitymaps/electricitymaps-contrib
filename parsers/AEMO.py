import io
import re
import zipfile
from datetime import datetime
from logging import Logger, getLogger
from typing import Any
from urllib.parse import urljoin
from zoneinfo import ZoneInfo

import pandas as pd
from bs4 import BeautifulSoup
from requests import Session

from electricitymap.contrib.lib.models.event_lists import TotalConsumptionList
from electricitymap.contrib.lib.models.events import EventSourceType
from electricitymap.contrib.lib.types import ZoneKey

SOURCE = "aemo.com.au"

ZONE_KEY_TO_REGION = {
    "AU-NSW": "NSW1",
    "AU-QLD": "QLD1",
    "AU-SA": "SA1",
    "AU-TAS": "TAS1",
    "AU-VIC": "VIC1",
    "AU-WA": "WEM",  # This zone is not implemented yet
}

ZONE_KEY_TO_TIMEZONE = {
    "AU-NSW": ZoneInfo("Australia/Sydney"),
    "AU-QLD": ZoneInfo("Australia/Brisbane"),
    "AU-SA": ZoneInfo("Australia/Adelaide"),
    "AU-TAS": ZoneInfo("Australia/Hobart"),
    "AU-VIC": ZoneInfo("Australia/Melbourne"),
    "AU-WA": ZoneInfo("Australia/Perth"),  # This zone is not implemented yet
}

# TODO, what about the other zone in Australia (AU-WA)? Check remaining zone


def find_document(session, target_datetime):
    # Fetch the directory listing
    base_url = "http://nemweb.com.au/Reports/CURRENT/Operational_Demand/FORECAST_HH/"
    response = session.get(base_url)

    # Parse with BeautifulSoup
    soup = BeautifulSoup(response.text, "html.parser")

    # String datetime
    target_date_str = target_datetime.strftime("%Y%m%d")

    # Find matching links
    matching_links = soup.find_all(
        "a",
        href=re.compile(
            rf"PUBLIC_FORECAST_OPERATIONAL_DEMAND_HH_{target_date_str}\d+_\d+\.zip"
        ),
    )

    if matching_links:
        target_link = matching_links[-1]  # Get the last (most recent) link
        full_url = urljoin(base_url, target_link["href"])  # Construct full URL

        file_response = session.get(full_url)

        with zipfile.ZipFile(io.BytesIO(file_response.content)) as z:
            csv_filename = z.namelist()[0]
            with z.open(csv_filename) as f:
                df = pd.read_csv(f)
                return df
    else:
        print("No matching files found")


def fetch_consumption_forecast(
    zone_key: ZoneKey,  # "AU-NSW", "AU-QLD", "AU-SA", "AU-TAS", "AU-VIC", "AU-WA"
    session: Session | None = None,
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
) -> list[dict[str, Any]]:
    """Consumption forecast in MW every half an hour for 10 days ahead.
    Only for NSW1, QND1, SA1, TAS1, VIC1 zones."""
    session = session or Session()

    if target_datetime is None:
        target_datetime = datetime.now(tz=ZONE_KEY_TO_TIMEZONE[zone_key])

    df = find_document(session, target_datetime)

    # Transform dataframe
    df.columns = df.iloc[0]
    df = df.iloc[1:-1].reset_index(drop=True)

    #
    region = ZONE_KEY_TO_REGION.get(zone_key)
    all_consumption_events = df[
        df["REGIONID"] == region
    ]  # all events with a datetime and a consumption value
    consumption_list = TotalConsumptionList(logger)
    for _, event in all_consumption_events.iterrows():
        datetime_object = datetime.strptime(
            event["INTERVAL_DATETIME"], "%Y/%m/%d %H:%M:%S"
        ).replace(tzinfo=ZONE_KEY_TO_TIMEZONE[zone_key])

        consumption_list.append(
            zoneKey=zone_key,
            datetime=datetime_object,
            consumption=float(
                event["OPERATIONAL_DEMAND_POE50"]
            ),  # 50% probability of exceedance operational demand forecast value
            source=SOURCE,
            sourceType=EventSourceType.forecasted,
        )
    return consumption_list.to_list()


if __name__ == "__main__":
    """Main method, never used by the electricityMap backend, but handy for testing."""

    print(fetch_consumption_forecast("AU-NSW"))
    print(fetch_consumption_forecast("AU-QLD"))
    print(fetch_consumption_forecast("AU-SA"))
    print(fetch_consumption_forecast("AU-TAS"))
    print(fetch_consumption_forecast("AU-VIC"))
    print(
        fetch_consumption_forecast("AU-WA")
    )  # Not implemented yet. It returns an empty list
