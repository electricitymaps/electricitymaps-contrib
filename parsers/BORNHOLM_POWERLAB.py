#!/usr/bin/env python3
from datetime import datetime
from logging import Logger, getLogger
from zoneinfo import ZoneInfo

from requests import Session

from electricitymap.contrib.lib.models.event_lists import (
    ExchangeList,
    ProductionBreakdownList,
)
from electricitymap.contrib.lib.models.events import ProductionMix
from electricitymap.contrib.lib.types import ZoneKey

PRODUCTION_MAPPING = {
    "wind": "wind_turbines",
    "biomass": "factory",
    "solar": "solar_cells",
}
LATEST_DATA_URL = "https://bornholm.powerlab.dk/visualizer/latestdata"

SOURCE = "bornholm.powerlab.dk"
TIMEZONE = ZoneInfo("UTC")


def fetch_production(
    zone_key: ZoneKey = ZoneKey("DK-BHM"),
    session: Session | None = None,
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
) -> list[dict]:
    """Requests the last known production mix (in MW) of a given country."""
    if target_datetime:
        raise NotImplementedError("This parser is not yet able to parse past dates")
    if session is None:
        session = Session()
    response = session.get(LATEST_DATA_URL).json()
    production = ProductionBreakdownList(logger)
    production.append(
        zoneKey=zone_key,
        datetime=datetime.fromtimestamp(response["latest"], tz=TIMEZONE),
        source=SOURCE,
        production=ProductionMix(
            biomass=response["sub"]["factory"],
            solar=response["sub"]["solar_cells"],
            wind=response["sub"]["wind_turbines"],
        ),
    )
    return production.to_list()


def fetch_exchange(
    zone_key1: ZoneKey = ZoneKey("DK-BHM"),
    zone_key2: ZoneKey = ZoneKey("SE-SE4"),
    session: Session | None = None,
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
) -> list[dict]:
    """Requests the last known power exchange (in MW) between two countries."""
    if target_datetime:
        raise NotImplementedError("This parser is not yet able to parse past dates")
    if session is None:
        session = Session()
    response = session.get(LATEST_DATA_URL).json()
    exchange = ExchangeList(logger)
    sorted_zone_keys = sorted([zone_key1, zone_key2])
    direction = 1 if zone_key1 == sorted_zone_keys[0] else -1
    exchange.append(
        zoneKey=ZoneKey("->".join(sorted_zone_keys)),
        datetime=datetime.fromtimestamp(response["latest"], tz=TIMEZONE),
        source=SOURCE,
        netFlow=direction * response["sub"]["seacable"],
    )

    return exchange.to_list()


if __name__ == "__main__":
    """Main method, never used by the Electricity Map backend, but handy for testing."""

    print("fetch_production() ->")
    print(fetch_production())
    print("fetch_exchange(DK-BHM, SE-SE4) ->")
    print(fetch_exchange())
