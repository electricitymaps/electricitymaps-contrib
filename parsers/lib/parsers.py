import importlib

from electricitymap.contrib.config import EXCHANGES_CONFIG, ZONES_CONFIG

# Prepare all parsers
CONSUMPTION_PARSERS = {}
PRODUCTION_PARSERS = {}
PRODUCTION_PER_MODE_FORECAST_PARSERS = {}
PRODUCTION_PER_UNIT_PARSERS = {}
EXCHANGE_PARSERS = {}
PRICE_PARSERS = {}
CONSUMPTION_FORECAST_PARSERS = {}
GENERATION_FORECAST_PARSERS = {}
EXCHANGE_FORECAST_PARSERS = {}
PRODUCTION_CAPACITY_PARSERS = {}
DAYAHEAD_LOCATIONAL_MARGINAL_PRICE_PARSERS = {}
REALTIME_LOCATIONAL_MARGINAL_PRICE_PARSERS = {}

PARSER_KEY_TO_DICT = {
    "consumption": CONSUMPTION_PARSERS,
    "production": PRODUCTION_PARSERS,
    "productionPerUnit": PRODUCTION_PER_UNIT_PARSERS,
    "productionPerModeForecast": PRODUCTION_PER_MODE_FORECAST_PARSERS,
    "exchange": EXCHANGE_PARSERS,
    "price": PRICE_PARSERS,
    "dayaheadLocationalMarginalPrice": DAYAHEAD_LOCATIONAL_MARGINAL_PRICE_PARSERS,
    "realtimeLocationalMarginalPrice": REALTIME_LOCATIONAL_MARGINAL_PRICE_PARSERS,
    "consumptionForecast": CONSUMPTION_FORECAST_PARSERS,
    "generationForecast": GENERATION_FORECAST_PARSERS,
    "exchangeForecast": EXCHANGE_FORECAST_PARSERS,
    "productionCapacity": PRODUCTION_CAPACITY_PARSERS,
}


def _parser_key_to_parser_folder(parser_key: str):
    return (
        "electricitymap.contrib.capacity_parsers"
        if parser_key == "productionCapacity"
        else "parsers"
    )


# Read all zones
for zone_id, zone_config in ZONES_CONFIG.items():
    for parser_key, v in zone_config.get("parsers", {}).items():
        mod_name, fun_name = v.split(".")
        mod = importlib.import_module(
            f"{_parser_key_to_parser_folder(parser_key)}.{mod_name}"
        )
        PARSER_KEY_TO_DICT[parser_key][zone_id] = getattr(mod, fun_name)


# Read all exchanges
for exchange_id, exchange_config in EXCHANGES_CONFIG.items():
    for parser_key, v in exchange_config.get("parsers", {}).items():
        mod_name, fun_name = v.split(".")
        mod = importlib.import_module(f"parsers.{mod_name}")
        PARSER_KEY_TO_DICT[parser_key][exchange_id] = getattr(mod, fun_name)
