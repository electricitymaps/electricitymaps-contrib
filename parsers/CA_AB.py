#!/usr/bin/env python3

"""Parse the Alberta Electric System Operator's (AESO's) Energy Trading System
(ETS) website.
"""

# Standard library imports
import csv
import re
import urllib.parse
from datetime import datetime, timedelta
from logging import Logger, getLogger
from typing import Any
from zoneinfo import ZoneInfo

# Third-party library imports
from requests import Session

from electricitymap.contrib.lib.models.event_lists import ExchangeList, PriceList
from electricitymap.contrib.lib.types import ZoneKey

# Local library imports
from parsers.lib import validation

DEFAULT_ZONE_KEY = ZoneKey("CA-AB")
MINIMUM_PRODUCTION_THRESHOLD = 10  # MW
TIMEZONE = ZoneInfo("Canada/Mountain")
URL = urllib.parse.urlsplit("http://ets.aeso.ca/ets_web/ip/Market/Reports")
URL_STRING = urllib.parse.urlunsplit(URL)


def fetch_exchange(
    zone_key1: str = DEFAULT_ZONE_KEY,
    zone_key2: str = "CA-BC",
    session: Session | None = None,
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
) -> list[dict[str, Any]]:
    """Request the last known power exchange (in MW) between two countries."""
    if target_datetime:
        raise NotImplementedError("Currently unable to scrape historical data")
    session = session or Session()
    response = session.get(
        f"{URL_STRING}/CSDReportServlet", params={"contentType": "csv"}
    )
    interchange = dict(csv.reader(response.text.split("\r\n\r\n")[4].splitlines()))
    flows = {
        f"{DEFAULT_ZONE_KEY}->CA-BC": interchange["British Columbia"],
        f"{DEFAULT_ZONE_KEY}->CA-SK": interchange["Saskatchewan"],
        f"{DEFAULT_ZONE_KEY}->US-MT": interchange["Montana"],
        f"{DEFAULT_ZONE_KEY}->US-NW-NWMT": interchange["Montana"],
    }
    sorted_zone_keys = ZoneKey("->".join(sorted((zone_key1, zone_key2))))
    if sorted_zone_keys not in flows:
        raise NotImplementedError(f"Pair '{sorted_zone_keys}' not implemented")
    exchanges = ExchangeList(logger)
    exchanges.append(
        zoneKey=sorted_zone_keys,
        datetime=get_csd_report_timestamp(response.text),
        netFlow=float(flows[sorted_zone_keys]),
        source=URL.netloc,
    )
    return exchanges.to_list()


def fetch_price(
    zone_key: ZoneKey = DEFAULT_ZONE_KEY,
    session: Session | None = None,
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
) -> list[dict[str, Any]]:
    """Request the last known power price of a given country."""
    if target_datetime:
        raise NotImplementedError("Currently unable to scrape historical data")
    session = session or Session()
    response = session.get(
        f"{URL_STRING}/SMPriceReportServlet", params={"contentType": "csv"}
    )
    prices = PriceList(logger)
    for row in csv.reader(response.text.split("\r\n\r\n")[2].splitlines()[1:]):
        if row[1] != "-":
            date, hour = row[0].split()
            prices.append(
                zoneKey=zone_key,
                datetime=datetime.strptime(
                    f"{date} {int(hour) - 1}", "%m/%d/%Y %H"
                ).replace(tzinfo=TIMEZONE)
                + timedelta(hours=1),
                price=float(row[1]),
                source=URL.netloc,
                currency="CAD",
            )
    return prices.to_list()


def fetch_production(
    zone_key: str = DEFAULT_ZONE_KEY,
    session: Session | None = None,
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
) -> dict[str, Any] | None:
    """Request the last known production mix (in MW) of a given country."""
    if target_datetime:
        raise NotImplementedError("This parser is not yet able to parse past dates")
    session = session or Session()
    response = session.get(
        f"{URL_STRING}/CSDReportServlet", params={"contentType": "csv"}
    )
    generation = {
        row[0]: {
            "MC": float(row[1]),  # maximum capability
            "TNG": float(row[2]),  # total net generation
        }
        for row in csv.reader(response.text.split("\r\n\r\n")[3].splitlines())
    }
    return validation.validate(
        {
            "capacity": {
                "gas": generation["COGENERATION"]["MC"]
                + generation["COMBINED CYCLE"]["MC"]
                + generation["GAS FIRED STEAM"]["MC"]
                + generation["SIMPLE CYCLE"]["MC"],
                "wind": generation["WIND"]["MC"],
                "solar": generation["SOLAR"]["MC"],
                "hydro": generation["HYDRO"]["MC"],
                "biomass": generation["OTHER"]["MC"],
                "battery storage": generation["ENERGY STORAGE"]["MC"],
            },
            "datetime": get_csd_report_timestamp(response.text),
            "production": {
                "gas": generation["COGENERATION"]["TNG"]
                + generation["COMBINED CYCLE"]["TNG"]
                + generation["GAS FIRED STEAM"]["TNG"]
                + generation["SIMPLE CYCLE"]["TNG"],
                "wind": generation["WIND"]["TNG"],
                "solar": generation["SOLAR"]["TNG"],
                "hydro": generation["HYDRO"]["TNG"],
                "biomass": generation["OTHER"]["TNG"],
            },
            "source": URL.netloc,
            "storage": {
                "battery": generation["ENERGY STORAGE"]["TNG"],
            },
            "zoneKey": zone_key,
        },
        logger,
        floor=MINIMUM_PRODUCTION_THRESHOLD,
        remove_negative=True,
    )


def get_csd_report_timestamp(report):
    """Get the timestamp from a current supply/demand (CSD) report."""
    return datetime.strptime(
        re.search(r'"Last Update : (.*)"', report).group(1), "%b %d, %Y %H:%M"
    ).replace(tzinfo=TIMEZONE)


if __name__ == "__main__":
    # Never used by the electricityMap backend, but handy for testing.
    print("fetch_production() ->")
    print(fetch_production())
    print("fetch_price() ->")
    print(fetch_price())
    print(f"fetch_exchange({DEFAULT_ZONE_KEY}, CA-BC) ->")
    print(fetch_exchange(DEFAULT_ZONE_KEY, "CA-BC"))
    print(f"fetch_exchange({DEFAULT_ZONE_KEY}, CA-SK) ->")
    print(fetch_exchange(DEFAULT_ZONE_KEY, "CA-SK"))
    print(f"fetch_exchange({DEFAULT_ZONE_KEY}, US-MT) ->")
    print(fetch_exchange(DEFAULT_ZONE_KEY, "US-MT"))
    print(f"fetch_exchange({DEFAULT_ZONE_KEY}, US-NW-NWMT) ->")
    print(fetch_exchange(DEFAULT_ZONE_KEY, "US-NW-NWMT"))
