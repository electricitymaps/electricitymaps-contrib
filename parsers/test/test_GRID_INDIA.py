from datetime import datetime
from zoneinfo import ZoneInfo

import pytest
from requests_mock import GET

from electricitymap.contrib.config import ZONE_NEIGHBOURS, ZoneKey
from parsers import GRID_INDIA


@pytest.mark.parametrize("neighbour_zone_key", ZONE_NEIGHBOURS[ZoneKey("IN-EA")])
def test_exchanges(adapter, session, snapshot, neighbour_zone_key: ZoneKey):
    target_date = datetime(2023, 6, 25, tzinfo=ZoneInfo("Asia/Kolkata"))

    pdfurl = GRID_INDIA.get_psp_report_file_url(target_date)

    with open("parsers/test/mocks/GRID_INDIA/exchanges.pdf", "rb") as data:
        adapter.register_uri(
            GET,
            pdfurl,
            content=data.read(),
        )
    assert snapshot == GRID_INDIA.fetch_exchange(
        ZoneKey("IN-EA"),
        neighbour_zone_key,
        session,
        datetime(2023, 6, 25, tzinfo=ZoneInfo("Asia/Kolkata")),
    )

