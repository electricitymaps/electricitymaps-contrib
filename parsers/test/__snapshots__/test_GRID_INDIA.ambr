# serializer version: 1
# name: test_exchanges[BD]
  list([
    dict({
      'datetime': datetime.datetime(2023, 6, 25, 0, 0, tzinfo=zoneinfo.ZoneInfo(key='Asia/Kolkata')),
      'netFlow': -936.667,
      'sortedZoneKeys': 'BD->IN-EA',
      'source': 'grid-india.in',
      'sourceType': <EventSourceType.measured: 'measured'>,
    }),
  ])
# ---
# name: test_exchanges[BT]
  list([
    dict({
      'datetime': datetime.datetime(2023, 6, 25, 0, 0, tzinfo=zoneinfo.ZoneInfo(key='Asia/Kolkata')),
      'netFlow': 731.25,
      'sortedZoneKeys': 'BT->IN-EA',
      'source': 'grid-india.in',
      'sourceType': <EventSourceType.measured: 'measured'>,
    }),
  ])
# ---
# name: test_exchanges[IN-NE]
  list([
    dict({
      'datetime': datetime.datetime(2023, 6, 25, 0, 0, tzinfo=zoneinfo.ZoneInfo(key='Asia/Kolkata')),
      'netFlow': -1150.0,
      'sortedZoneKeys': 'IN-EA->IN-NE',
      'source': 'grid-india.in',
      'sourceType': <EventSourceType.measured: 'measured'>,
    }),
  ])
# ---
# name: test_exchanges[IN-NO]
  list([
    dict({
      'datetime': datetime.datetime(2023, 6, 25, 0, 0, tzinfo=zoneinfo.ZoneInfo(key='Asia/Kolkata')),
      'netFlow': -1637.5,
      'sortedZoneKeys': 'IN-EA->IN-NO',
      'source': 'grid-india.in',
      'sourceType': <EventSourceType.measured: 'measured'>,
    }),
  ])
# ---
# name: test_exchanges[IN-SO]
  list([
    dict({
      'datetime': datetime.datetime(2023, 6, 25, 0, 0, tzinfo=zoneinfo.ZoneInfo(key='Asia/Kolkata')),
      'netFlow': -3083.333,
      'sortedZoneKeys': 'IN-EA->IN-SO',
      'source': 'grid-india.in',
      'sourceType': <EventSourceType.measured: 'measured'>,
    }),
  ])
# ---
# name: test_exchanges[IN-WE]
  list([
    dict({
      'datetime': datetime.datetime(2023, 6, 25, 0, 0, tzinfo=zoneinfo.ZoneInfo(key='Asia/Kolkata')),
      'netFlow': 2029.167,
      'sortedZoneKeys': 'IN-EA->IN-WE',
      'source': 'grid-india.in',
      'sourceType': <EventSourceType.measured: 'measured'>,
    }),
  ])
# ---
# name: test_exchanges[NP]
  list([
    dict({
      'datetime': datetime.datetime(2023, 6, 25, 0, 0, tzinfo=zoneinfo.ZoneInfo(key='Asia/Kolkata')),
      'netFlow': -253.333,
      'sortedZoneKeys': 'IN-EA->NP',
      'source': 'grid-india.in',
      'sourceType': <EventSourceType.measured: 'measured'>,
    }),
  ])
# ---
